packages/artifact/src/upload-http-client.ts

```diff
@@ -0,0 +1,326 @@
+          parameters.resourceUrl,
+          chunk,
+          start,
+          end,
+          fileSize
+        )
+
+        if (!result) {
+          /**
+           * Chunk failed to upload, report as failed and do not continue uploading any more chunks for the file. It is possible that part of a chunk was
+           * successfully uploaded so the server may report a different size for what was uploaded
+           **/
+          isUploadSuccessful = false
+          failedChunkSizes += chunkSize
+          warning(`Aborting upload for ${parameters.file} due to failure`)
+          abortFileUpload = true
+        }
+      }
+    })
+  )
+  return {
+    isSuccess: isUploadSuccessful,
+    successfulUploadSize: fileSize - failedChunkSizes
+  }
+}
+
+/**
+ * Uploads a chunk of an individual file to the specified resourceUrl. If the upload fails and the status code
+ * indicates a retryable status, we try to upload the chunk as well
+ * @param {HttpClient} restClient RestClient that will be making the appropriate HTTP call
+ * @param {string} resourceUrl Url of the resource that the chunk will be uploaded to
+ * @param {NodeJS.ReadableStream} data Stream of the file that will be uploaded
+ * @param {number} start Starting byte index of file that the chunk belongs to
+ * @param {number} end Ending byte index of file that the chunk belongs to
+ * @param {number} totalSize Total size of the file in bytes that is being uploaded
+ * @returns if the chunk was successfully uploaded
+ */
+async function uploadChunk(
+  restClient: HttpClient,
+  resourceUrl: string,
+  data: NodeJS.ReadableStream,
+  start: number,
+  end: number,
+  totalSize: number
+): Promise<boolean> {
+  // eslint-disable-next-line no-console
+  console.log(
+    `Uploading chunk of size ${end -
+      start +
+      1} bytes at offset ${start} with content range: ${getContentRange(
+      start,
+      end,
+      totalSize
+    )}`
+  )
+
+  const requestOptions = getRequestOptions(
+    'application/octet-stream',
+    totalSize,
+    getContentRange(start, end, totalSize)
+  )
+
+  const uploadChunkRequest = async (): Promise<IHttpClientResponse> => {
+    return await restClient.sendStream('PUT', resourceUrl, data, requestOptions)
+  }
+
+  const response = await uploadChunkRequest()
+  if (isSuccessStatusCode(response.message.statusCode)) {
+    debug(
+      `Chunk for ${start}:${end} was successfully uploaded to ${resourceUrl}`
+    )
+    return true
+  } else if (isRetryableStatusCode(response.message.statusCode)) {
+    // eslint-disable-next-line no-console
+    console.log(
+      `Received http ${response.message.statusCode} during chunk upload, will retry at offset ${start} after 10 seconds.`
+    )
+    await new Promise(resolve => setTimeout(resolve, 10000))
+    const retryResponse = await uploadChunkRequest()
+    if (isSuccessStatusCode(retryResponse.message.statusCode)) {
+      return true
+    } else {
+      // eslint-disable-next-line no-console
```

Comment:
i think stdout is preferred so unit tests can be less noisy

e.g. here is a snippet from somewhere else in toolkit
```js
        process.stdout.write(`GITHUB_EVENT_PATH ${path} does not exist${EOL}`)
```

Reply:
not sure, the linter may have another reason for avoiding console.log

Reply:
Perhaps we could disable console.log in the jest configuration? https://stackoverflow.com/questions/44467657/jest-better-way-to-disable-console-inside-unit-tests

Some of the response logging that I have is really noisy, for the tests I'll just mock `console.log` so that there is no annoying output

Reply:
This was a suggestion for handling commands in tests https://github.com/actions/toolkit/issues/118#issuecomment-528888332 and it was later added to https://github.com/actions/cache/blob/master/jest.config.js#L15-L23. I add this to all of my jest configs now. It might be good to add to the TS template.

Reply:
I ended up doing the following for each of the tests
 ```
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(core, 'debug').mockImplementation(() => {})
    jest.spyOn(core, 'info').mockImplementation(() => {})
    jest.spyOn(core, 'warning').mockImplementation(() => {})
```
Shouldn't be any noise in the tests now